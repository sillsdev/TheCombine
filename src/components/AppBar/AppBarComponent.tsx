import React from "react";
import AppBar from "@material-ui/core/AppBar";
import Toolbar from "@material-ui/core/Toolbar";
import { withLocalize, LocalizeContextProps } from "react-localize-redux";
import { Grid } from "@material-ui/core";
import UserMenu from "./UserMenu";
import GoToHomeButton from "../GoToHomeButton/GoToHomeButtonComponent";
import theme from "../../types/theme";

/** An app bar shown at the top of almost every page of The Combine */
export class AppBarComponent extends React.Component<LocalizeContextProps> {
  render() {
    return (
      <div className="NavigationBar" style={{ marginBottom: 64 }}>
        <AppBar position="fixed" style={{ zIndex: theme.zIndex.drawer + 1 }}>
          <Toolbar>
            <Grid
              container
              justify="space-between"
              spacing={2}
              alignItems="center"
            >
              <Grid item>
                <GoToHomeButton />
              </Grid>
              <Grid item>
                <UserMenu />
              </Grid>
            </Grid>
          </Toolbar>
        </AppBar>
      </div>
    );
  }
}

export default withLocalize(AppBarComponent);
