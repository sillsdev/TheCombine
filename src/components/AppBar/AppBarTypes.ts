import { Path } from "types/path";
import { themeColors } from "types/theme";

export const buttonMinHeight = 40;
export const appBarHeight = 64;

export interface TabProps {
  currentTab: Path;
}

export function tabColor(currentTab: Path, tabName: Path): string {
  return currentTab.indexOf(tabName) !== -1
    ? themeColors.darkShade
    : themeColors.lightShade;
}

export function shortenName(name: string, maxLength: number): string {
  if (maxLength < 7) {
    maxLength = 7;
    console.warn("shortenName() cannot be used with maxLength < 7.");
  }
  if (name.length <= maxLength) {
    return name;
  }
  const halfLength = (maxLength - 3) / 2;
  const front = name.substring(0, halfLength);
  const back = name.substring(name.length - halfLength);
  return `${front}...${back}`;
}
