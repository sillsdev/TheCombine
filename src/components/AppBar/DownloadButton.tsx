import { IconButton, Tooltip } from "@material-ui/core";
import { GetApp } from "@material-ui/icons";
import { HubConnection, HubConnectionBuilder } from "@microsoft/signalr";
import React, { useEffect } from "react";
import { Translate } from "react-localize-redux";

/** An app bar shown at the top of almost every page of The Combine */
export function DownloadButton() {
  const [connection, setConnection] = React.useState<null | HubConnection>(
    null
  );
  const [received, setReceived] = React.useState<string>("");

  useEffect(() => {
    const newConnection = new HubConnectionBuilder()
      .withUrl("https://localhost:5001/queue")
      .withAutomaticReconnect()
      .build();
    setConnection(newConnection);
  }, []);

  useEffect(() => {
    if (connection) {
      connection
        .start()
        .then(() => {
          console.log("Connected!");
          connection.on("ReceiveMessage", (user: string, message: string) => {
            setReceived(`${user}: ${message}`);
          });
        })
        .catch((e) => console.log("Connection failed: ", e));
    }
  }, [connection]);

  function download() {
    setReceived("");
  }

  return (
      <React.Fragment>
       {received &&
      <Tooltip title={<Translate id="appBar.downloadReady" />} placement="bottom">
        <IconButton tabIndex={-1} onClick={download}>
          <GetApp />
        </IconButton>
      </Tooltip>}
      </React.Fragment>
    );
  };

export default React.memo<{}>(DownloadButton);