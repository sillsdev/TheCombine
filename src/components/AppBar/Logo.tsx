import React from "react";
import { Button } from "@material-ui/core";
import logo from "../../resources/CombineLogoV1.png";

/** A button that redirects to the home page */
export default function Logo() {
  return (
    <Button>
      <img src={logo} height="50" alt="Logo" />
    </Button>
  );
}
