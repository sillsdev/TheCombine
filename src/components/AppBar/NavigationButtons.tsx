import React from "react";
import { Button } from "@material-ui/core";
import history from "../../history";
import { Translate } from "react-localize-redux";

/** A button that redirects to the home page */
export default function NavigationButtons() {
  return (
    <React.Fragment>
      <Button
        onClick={() => {
          history.location.pathname === "/"
            ? history.push("/")
            : history.push("/data-entry");
        }}
        color="inherit"
      >
        <Translate id="appBar.dataEntry" />
      </Button>
      <Button
        onClick={() => {
          history.location.pathname === "/"
            ? history.push("/")
            : history.push("/goals");
        }}
        color="inherit"
      >
        <Translate id="appBar.dataCleanup" />
      </Button>
    </React.Fragment>
  );
}
