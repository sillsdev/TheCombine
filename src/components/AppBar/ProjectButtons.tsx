import { BarChart, Settings } from "@mui/icons-material";
import { Button, Hidden, Tooltip, Typography } from "@mui/material";
import { ReactElement, useEffect, useState } from "react";
import { useTranslation } from "react-i18next";
import { useSelector } from "react-redux";
import { useNavigate } from "react-router-dom";

import { Permission } from "api/models";
import { getCurrentPermissions } from "backend";
import { getCurrentUser } from "backend/localStorage";
import {
  TabProps,
  buttonMinHeight,
  shortenName,
  tabColor,
} from "components/AppBar/AppBarTypes";
import { StoreState } from "types";
import { Path } from "types/path";

export const projButtonId = "project-settings";
export const statButtonId = "project-statistics";

const enum projNameLength {
  sm = 15,
  md = 25,
  lg = 45,
  xl = 75,
}

export async function getHasStatsPermission(): Promise<boolean> {
  if (getCurrentUser()?.isAdmin) {
    return true;
  }
  return (await getCurrentPermissions()).includes(Permission.Statistics);
}

/** A button that redirects to the project settings */
export default function ProjectButtons(props: TabProps): ReactElement {
  const projectName = useSelector(
    (state: StoreState) => state.currentProjectState.project.name
  );
  const [hasStatsPermission, setHasStatsPermission] = useState<boolean>(false);
  const { t } = useTranslation();
  const navigate = useNavigate();

  useEffect(() => {
    getHasStatsPermission().then(setHasStatsPermission);
  }, []);

  return (
    <>
      {hasStatsPermission && (
        <Tooltip title={t("appBar.statistics")}>
          <Button
            id={statButtonId}
            onClick={() => navigate(Path.Statistics)}
            color="inherit"
            style={{
              background: tabColor(props.currentTab, Path.Statistics),
              minHeight: buttonMinHeight,
              minWidth: 0,
              margin: 5,
            }}
          >
            <BarChart />
          </Button>
        </Tooltip>
      )}
      <Tooltip title={t("appBar.projectSettings")}>
        <Button
          id={projButtonId}
          onClick={() => navigate(Path.ProjSettings)}
          color="inherit"
          style={{
            background: tabColor(props.currentTab, Path.ProjSettings),
            minHeight: buttonMinHeight,
            minWidth: 0,
          }}
        >
          <Settings />
          <Hidden smDown>
            <Typography
              display="inline"
              style={{ marginLeft: 5, marginRight: 5 }}
            >
              <Hidden xlDown>
                {shortenName(projectName, projNameLength.xl)}
              </Hidden>
              <Hidden lgDown xlUp>
                {shortenName(projectName, projNameLength.lg)}
              </Hidden>
              <Hidden mdDown lgUp>
                {shortenName(projectName, projNameLength.md)}
              </Hidden>
              <Hidden mdUp smDown>
                {shortenName(projectName, projNameLength.sm)}
              </Hidden>
            </Typography>
          </Hidden>
        </Button>
      </Tooltip>
    </>
  );
}
