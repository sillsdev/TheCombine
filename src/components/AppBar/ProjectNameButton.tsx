import { Button, Hidden, Tooltip } from "@material-ui/core";
import { Settings } from "@material-ui/icons";
import { Translate } from "react-localize-redux";
import { useSelector } from "react-redux";

import history, { Path } from "browserHistory";
import { StoreState } from "types";
import { tabColor } from "types/theme";

interface ProjectNameButtonProps {
  currentTab: Path;
}

/** A button that redirects to the project settings */
export default function ProjectNameButton(props: ProjectNameButtonProps) {
  const projectName = useSelector(
    (state: StoreState) => state.currentProjectState.project.name
  );
  const background = tabColor(props.currentTab, Path.ProjSettings);
  return (
    <Tooltip title={<Translate id="userMenu.projectSettings" />}>
      <Button
        id="project-name"
        onClick={() => history.push(Path.ProjSettings)}
        color="inherit"
        style={{ background }}
      >
        <Settings />
        <Hidden xsDown>{projectName}</Hidden>
      </Button>
    </Tooltip>
  );
}
