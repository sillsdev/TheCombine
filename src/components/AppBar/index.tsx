import { AppBar, Stack, Toolbar } from "@mui/material";
import { ReactElement, useEffect, useState } from "react";
import { useLocation } from "react-router";

import { getProjectId } from "backend/localStorage";
import { appBarHeight } from "components/AppBar/AppBarTypes";
import Logo from "components/AppBar/Logo";
import NavigationButtons from "components/AppBar/NavigationButtons";
import ProjectButtons from "components/AppBar/ProjectButtons";
import UserGuideButton from "components/AppBar/UserGuideButton";
import UserMenu from "components/AppBar/UserMenu";
import { topBarHeight } from "components/LandingPage/TopBar";
import DownloadButton from "components/ProjectExport/DownloadButton";
import { Path } from "types/path";
import theme from "types/theme";
import { getBasePath } from "utilities/pathUtilities";

/** An app bar shown at the top of all logged in pages */
export default function AppBarComponent(): ReactElement {
  const location = useLocation();
  const [currentTab, setCurrentTab] = useState<Path>(Path.ProjScreen);

  useEffect(() => setCurrentTab(getBasePath(location.pathname)), [location]);

  return (
    <div className="NavigationBar" style={{ marginBottom: topBarHeight }}>
      <AppBar
        position="fixed"
        style={{ maxHeight: appBarHeight, zIndex: theme.zIndex.drawer + 1 }}
      >
        <Toolbar>
          <Stack
            alignItems="center"
            direction="row"
            justifyContent="space-between"
            width="100%"
          >
            <div>
              <Logo />
              {!!getProjectId() && (
                <NavigationButtons currentTab={currentTab} />
              )}
            </div>

            <div>
              {!!getProjectId() && <ProjectButtons currentTab={currentTab} />}
              <DownloadButton colorSecondary />
            </div>

            <div>
              <UserMenu currentTab={currentTab} />
              <UserGuideButton />
            </div>
          </Stack>
        </Toolbar>
      </AppBar>
    </div>
  );
}
