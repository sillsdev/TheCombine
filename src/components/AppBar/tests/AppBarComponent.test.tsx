import { Provider } from "react-redux";
import renderer, { ReactTestRenderer } from "react-test-renderer";
import configureMockStore from "redux-mock-store";

import { Path } from "browserHistory";
import { defaultState } from "components/App/DefaultState";
import AppBar from "components/AppBar/AppBarComponent";
import NavigationButtons from "components/AppBar/NavigationButtons";
import ProjectNameButton from "components/AppBar/ProjectNameButton";

const mockPath = jest.fn();
jest.mock("react-router-dom", () => ({
  useLocation: () => ({ pathname: mockPath() }),
}));

const mockStore = configureMockStore()(defaultState);

let testRenderer: ReactTestRenderer;

describe("AppBar", () => {
  it("renders without crashing", () => {
    mockPath.mockReturnValue(Path.ProjScreen);
    renderer.act(() => {
      testRenderer = renderer.create(
        <Provider store={mockStore}>
          <AppBar />
        </Provider>
      );
    });
  });
});

describe("NavigationButtons", () => {
  it("has only one tab shaded", () => {
    testRenderer = renderer.create(
      <Provider store={mockStore}>
        <NavigationButtons currentTab={Path.DataEntry} />
      </Provider>
    );
    expect(testRenderer.toJSON()).toMatchSnapshot();

    testRenderer = renderer.create(
      <Provider store={mockStore}>
        <ProjectNameButton currentTab={Path.Goals} />
      </Provider>
    );
    expect(testRenderer.toJSON()).toMatchSnapshot();
  });
});

describe("ProjectNameButton", () => {
  it("has tab shaded when itself is called", () => {
    mockPath.mockReturnValue(Path.ProjSettings);
    testRenderer = renderer.create(
      <Provider store={mockStore}>
        <ProjectNameButton currentTab={Path.ProjSettings} />
      </Provider>
    );
    expect(testRenderer.toJSON()).toMatchSnapshot();
  });
});
