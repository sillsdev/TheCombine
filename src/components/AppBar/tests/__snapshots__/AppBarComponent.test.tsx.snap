// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests AppBarComponent Ensures ProjectName Tab is shaded when itself is called 1`] = `
<button
  className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-colorInherit"
  disabled={false}
  id="project-name"
  onBlur={[Function]}
  onClick={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  style={
    Object {
      "background": "#1976d2",
    }
  }
  tabIndex={0}
  type="button"
>
  <span
    className="MuiButton-label"
  >
    
  </span>
</button>
`;

exports[`Tests AppBarComponent Ensures only one tab is shaded 1`] = `
Array [
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-colorInherit"
    disabled={false}
    id="data-entry"
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    style={
      Object {
        "background": "inherit",
      }
    }
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiButton-label"
    >
      Missing translationId: appBar.dataEntry for language: \${ languageCode }
    </span>
  </button>,
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-colorInherit"
    disabled={false}
    id="goals"
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    style={
      Object {
        "background": "#1976d2",
      }
    }
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiButton-label"
    >
      Missing translationId: appBar.dataCleanup for language: \${ languageCode }
    </span>
  </button>,
]
`;

exports[`Tests AppBarComponent Ensures only one tab is shaded 2`] = `
<button
  className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-colorInherit"
  disabled={false}
  id="project-name"
  onBlur={[Function]}
  onClick={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  style={
    Object {
      "background": "inherit",
    }
  }
  tabIndex={0}
  type="button"
>
  <span
    className="MuiButton-label"
  >
    
  </span>
</button>
`;
